### ISSUE TITLE
Support for Perl v5.16

### ISSUE THREAD

**Issue by johannessen** [majensen/perlbolt#44]
According to [p5p (PSC #025 discussion)](https://www.nntp.perl.org/group/perl.perl5.porters/2021/07/msg260781.html), around 15 % of Perl users are still on v5.16, due to this being the version that comes with RHEL7, which is expected to be supported until at least 2024.

@majensen: Do we want to target this version?

Neo4j::Bolt currently [fails](http://matrix.cpantesters.org/?dist=Neo4j-Bolt+0.4201) on Perl v5.16 and earlier. The CPAN testers reports suggest `av_top_index()` is the culprit. This function was only [introduced](https://metacpan.org/release/RJBS/perl-5.18.0/changes#Internal-Changes) in Perl v5.18. Replacing these with `av_len()` may or may not solve the problem (I haven’t had a chance to actually try this yet, and there could be more trouble that hasn’t shown up on CPAN testers just yet).

https://github.com/majensen/perlbolt/blob/3a3c14dc3d8806f089c02d85eec3c3858d5d66ac/lib/Neo4j/Bolt/CTypeHandlers.xs#L151

https://github.com/majensen/perlbolt/blob/3a3c14dc3d8806f089c02d85eec3c3858d5d66ac/lib/Neo4j/Bolt/CTypeHandlers.xs#L208

**Comment by majensen** [comment]
@johannessen you have the eagle eye. I will have a look.

**Comment by provector** [comment]
Hi Guys, I can confirm that the above change fixed the installation issue with perl 5.16 (PR created)
```
--> Working on Neo4j-Bolt-0.4201.tar.gz
Fetching file:///home/mpiekarski/Neo4j-Bolt-0.4201.tar.gz ... OK
Configuring Neo4j-Bolt-0.4201 ... OK
Building and testing Neo4j-Bolt-0.4201 ... OK
Successfully installed Neo4j-Bolt-0.4201
1 distribution installed
```
That's the good news :) The bad news is: it doesn't seem to work with Neo4j 4.3.2 :(
```
perl testDriver.pl 
ERROR: Bolt error -14: Could not agree on a protocol version at testDriver.pl line 37.
```
Any pointers on where should I even start to look (for issue, and possible solutions) would be greatly appreaciated!

Kind Regards

**Comment by provector** [comment]
I have actually seen your fork of neo4jlib with an open PR and support for bolt 4.0. I just managed to pull the master and compile it. I'm trying to understand now, how exactly the driver is dependent on the library (so I can merge the changes).

**Comment by johannessen** [comment]
Great that you got it to compile! But then it really *should* work with Neo4j 4.3 – unless you have an old version of the lib? But Neo4j::Client is supposed to take care of that. Weird…

**Comment by provector** [comment]
Yeah, I kinda figured the Neo4j::Client had your pr changes already in :) As for the above, I'm not sure that's the case anymore. Have a look at the latest compatibility table from Neo4j:
![image](https://user-images.githubusercontent.com/12935423/132531919-0eae7994-9f19-4cb0-ae12-b46a55ca64f9.png)

Following that document: https://7687.org/ it looks like 4.1 and 4.3 updated message specifications. Wouldn't that possibly break things?

**Comment by provector** [comment]
It seems like the neo4j-client itself works with neo 4.3.2, could it be something with Neo4j::Client wrapper? Where would I even start debugging that?

**Comment by johannessen** [comment]
Turns out the change suggested above isn’t enough to fix things for v5.16. We also use `is_utf8_invariant_string()`, which could be trickier to replace. I’ll look into it some time.

https://github.com/majensen/perlbolt/blob/3a3c14dc3d8806f089c02d85eec3c3858d5d66ac/lib/Neo4j/Bolt/CTypeHandlers.xs#L378

https://github.com/Perl/perl5/blob/034242a8a539b024648b18318271eabddf40ca48/inline.h#L510-L517

**Comment by johannessen** [comment]
I guess “some time” turned out to be right now.

For future reference, these are the relevant perl5 commits for `is_utf8_invariant_string()`: https://github.com/Perl/perl5/commit/eaf7a4d2ee7805b14e26e634fba0893913924a6c, https://github.com/Perl/perl5/commit/8871a094916837597c6a703c2e1b2306314df4fa

@majensen: Could you please take care of the release to both CPAN and GitHub?

@provector: I expect if you install the to-be-released version over whatever you have, you’ll get Bolt working. You *may* need to force-reinstall the lib (`cpanm -f Neo4j::Client`).

**Comment by johannessen** [comment]
According to [CPAN Testers](http://matrix.cpantesters.org/?dist=Neo4j-Bolt+0.4203), this was resolved in release [0.4203](https://metacpan.org/release/MAJENSEN/Neo4j-Bolt-0.4203).

**Comment by johannessen** [comment]
Addendum for future reference:
[ppport.h](https://metacpan.org/pod/Devel::PPPort), which I only just learned about, might offer a better solution for this kind of problem.
